server/container/util-java8/src/test/java/org/apache/james/util/CommutativityCheckerTest.java

```diff
@@ -0,0 +1,101 @@
+
+public class CommutativityCheckerTest {
+    @Test
+    public void constructorShouldThrowWhenNullValuesToTest() throws Exception {
+        BinaryOperator<Integer> binaryOperator = (a, b) -> a * a + b;
+
+        assertThatThrownBy(() -> new CommutativityChecker<>(null, binaryOperator))
+            .isInstanceOf(NullPointerException.class);
+    }
+
+    @Test
+    public void constructorShouldThrowWhenEmptyValuesToTest() throws Exception {
+        BinaryOperator<Integer> binaryOperator = (a, b) -> a * a + b;
+
+        assertThatThrownBy(() -> new CommutativityChecker<>(ImmutableSet.of(), binaryOperator))
+            .isInstanceOf(IllegalArgumentException.class);
+    }
+
+    @Test
+    public void constructorShouldThrowWhenSingleValueToTest() throws Exception {
+        BinaryOperator<Integer> binaryOperator = (a, b) -> a * a + b;
+
+        assertThatThrownBy(() -> new CommutativityChecker<>(ImmutableSet.of(0), binaryOperator))
+            .isInstanceOf(IllegalArgumentException.class);
+    }
+
+    @Test
+    public void constructorShouldThrowWhenNullOperation() throws Exception {
+        assertThatThrownBy(() -> new CommutativityChecker<>(ImmutableSet.of(0, 1), null))
+            .isInstanceOf(NullPointerException.class);
+    }
+
+    @Test
+    public void findNonCommutativeInputShouldReturnEmptyWhenCommutativeOperation() throws Exception {
+        Set<Integer> integers = ImmutableSet.of(5, 4, 3, 2, 1);
+        BinaryOperator<Integer> binaryOperator = (a, b) -> a + b;
+        CommutativityChecker<Integer> commutativityChecker = new CommutativityChecker<>(integers, binaryOperator);
+
+        assertThat(commutativityChecker.findNonCommutativeInput()).isEmpty();
+    }
+
+    @Test
+    public void findNonCommutativeInputShouldReturnDataWhenNonCommutativeOperation() throws Exception {
+        Set<Integer> integers = ImmutableSet.of(2, 1);
+        BinaryOperator<Integer> binaryOperator = (a, b) -> 2 * a + b;
+        CommutativityChecker<Integer> commutativityChecker = new CommutativityChecker<>(integers, binaryOperator);
+
+        assertThat(commutativityChecker.findNonCommutativeInput())
+            .containsOnly(Pair.of(2, 1));
+    }
+
+    @Test
+    public void findNonCommutativeInputShouldNotReturnDataWhenStable() throws Exception {
+        Set<Integer> integers = ImmutableSet.of(0, 1, 2);
+        BinaryOperator<Integer> binaryOperator = (a, b) -> a * a + b;
+        CommutativityChecker<Integer> commutativityChecker = new CommutativityChecker<>(integers, binaryOperator);
+
+        assertThat(commutativityChecker.findNonCommutativeInput())
+            .containsOnly(Pair.of(1, 2),
+                Pair.of(0, 2));
+    }
+
+    @Test
+    public void findNonCommutativeInputShouldReturnEmptyWhenNonCommutativeOperationButOnlyStableValues() throws Exception {
+        Set<Integer> integers = ImmutableSet.of(0, 1);
+        BinaryOperator<Integer> binaryOperator = (a, b) -> a * a + b;
```

Comment:
We can make this test more understandable by giving a better name to "binaryOperator"

Here I propose **nonCommutativeOperatorWithStableValues**

We can also call integres **stableValues**